File: src/envelope.ts
/**
 * Envelope types for MVP command request/response handling
 */

/** HeaderPrefix is the prefix for custom MVP headers in HTTP transport */
export const HEADER_PREFIX = 'x-mvp-';

/**
 * ErrorInfo provides structured error information
 */
export interface ErrorInfo {
  code: string;
  message: string;
}

/**
 * CmdReq represents a command request with headers
 */
export interface CmdReq {
  /** Cmd is the command name */
  cmd: string;
  /** Headers contains request metadata (auth tokens, trace IDs, etc.) */
  headers: Record<string, string>;
  /** Payload is the encoded command bytes/data */
  payload: unknown;
}

/**
 * CmdResp represents a command response with headers
 */
export interface CmdResp {
  /** Headers contains response metadata (new tokens, rate limits, etc.) */
  headers: Record<string, string>;
  /** Payload is the encoded result */
  payload: unknown;
  /** Error contains error information if the command failed */
  error?: ErrorInfo;
}

/**
 * Creates a new CmdReq with initialized headers map
 */
export function newCmdReq(cmd: string, payload?: unknown): CmdReq {
  return {
    cmd,
    headers: {},
    payload,
  };
}

/**
 * Creates a new CmdResp with initialized headers map
 */
export function newCmdResp(payload?: unknown): CmdResp {
  return {
    headers: {},
    payload,
  };
}

/**
 * Creates a new CmdResp with an error
 */
export function newCmdRespError(code: string, message: string): CmdResp {
  return {
    headers: {},
    payload: undefined,
    error: { code, message },
  };
}

/**
 * Helper class for building CmdReq with method chaining
 */
export class CmdReqBuilder {
  private req: CmdReq;

  constructor(cmd: string, payload?: unknown) {
    this.req = newCmdReq(cmd, payload);
  }

  /**
   * Adds a header to the request
   */
  withHeader(key: string, value: string): this {
    this.req.headers[key] = value;
    return this;
  }

  /**
   * Adds an authorization header
   */
  withAuth(token: string): this {
    return this.withHeader('auth', token);
  }

  /**
   * Sets the payload
   */
  withPayload(payload: unknown): this {
    this.req.payload = payload;
    return this;
  }

  /**
   * Builds and returns the CmdReq
   */
  build(): CmdReq {
    return this.req;
  }
}

/**
 * Helper class for building CmdResp with method chaining
 */
export class CmdRespBuilder {
  private resp: CmdResp;

  constructor(payload?: unknown) {
    this.resp = newCmdResp(payload);
  }

  /**
   * Adds a header to the response
   */
  withHeader(key: string, value: string): this {
    this.resp.headers[key] = value;
    return this;
  }

  /**
   * Sets an error on the response
   */
  withError(code: string, message: string): this {
    this.resp.error = { code, message };
    return this;
  }

  /**
   * Builds and returns the CmdResp
   */
  build(): CmdResp {
    return this.resp;
  }
}

/**
 * Checks if a response has an error
 */
export function hasError(resp: CmdResp): boolean {
  return resp.error !== undefined && resp.error !== null;
}

/**
 * Gets a header value from the response
 */
export function getHeader(resp: CmdResp, key: string): string | undefined {
  return resp.headers[key];
}

/**
 * Context type for passing request/response through async operations
 */
export interface MvpContext {
  /** Request being processed */
  request?: CmdReq;
  /** Response being built */
  response?: CmdResp;
  /** User data from authentication */
  user?: unknown;
  /** Request ID */
  requestId?: string;
  /** Additional context values */
  values: Record<string, unknown>;
}

/**
 * Creates a new empty context
 */
export function newContext(): MvpContext {
  return {
    values: {},
  };
}

/**
 * Creates a context with a request
 */
export function contextWithRequest(ctx: MvpContext, req: CmdReq): MvpContext {
  return {
    ...ctx,
    request: req,
  };
}

/**
 * Creates a context with a response
 */
export function contextWithResponse(ctx: MvpContext, resp: CmdResp): MvpContext {
  return {
    ...ctx,
    response: resp,
  };
}

/**
 * Creates a context with user data
 */
export function contextWithUser(ctx: MvpContext, user: unknown): MvpContext {
  return {
    ...ctx,
    user,
  };
}

/**
 * Gets the request header from context
 */
export function getRequestHeader(ctx: MvpContext, key: string): string | undefined {
  return ctx.request?.headers[key];
}

/**
 * Sets a response header in context (mutates the context response)
 */
export function setResponseHeader(ctx: MvpContext, key: string, value: string): MvpContext {
  if (!ctx.response) {
    ctx.response = newCmdResp();
  }
  ctx.response.headers[key] = value;
  return ctx;
}

File: runtime/go/mvp/envelope.go
package mvp

import "context"

// HeaderPrefix is the prefix for custom MVP headers in HTTP transport
const HeaderPrefix = "x-mvp-"

// Context keys for passing CmdReq/CmdResp through context
type cmdReqContextKey struct{}
type cmdRespContextKey struct{}

// CmdReq represents a command request with headers
type CmdReq struct {
	// Cmd is the command name
	Cmd string
	// Headers contains request metadata (auth tokens, trace IDs, etc.)
	Headers map[string]string
	// Payload is the encoded command bytes
	Payload []byte
}

// CmdResp represents a command response with headers
type CmdResp struct {
	// Headers contains response metadata (new tokens, rate limits, etc.)
	Headers map[string]string
	// Payload is the encoded result bytes
	Payload []byte
	// Error contains error information if the command failed
	Error *ErrorInfo
}

// ErrorInfo provides structured error information
type ErrorInfo struct {
	Code    string
	Message string
}

// NewCmdReq creates a new CmdReq with initialized headers map
func NewCmdReq(cmd string, payload []byte) *CmdReq {
	return &CmdReq{
		Cmd:     cmd,
		Headers: make(map[string]string),
		Payload: payload,
	}
}

// WithHeader adds a header to the request and returns the request for chaining
func (r *CmdReq) WithHeader(key, value string) *CmdReq {
	if r.Headers == nil {
		r.Headers = make(map[string]string)
	}
	r.Headers[key] = value
	return r
}

// WithAuth adds an authorization header
func (r *CmdReq) WithAuth(token string) *CmdReq {
	return r.WithHeader("auth", token)
}

// NewCmdResp creates a new CmdResp with initialized headers map
func NewCmdResp(payload []byte) *CmdResp {
	return &CmdResp{
		Headers: make(map[string]string),
		Payload: payload,
	}
}

// NewCmdRespError creates a new CmdResp with an error
func NewCmdRespError(code, message string) *CmdResp {
	return &CmdResp{
		Headers: make(map[string]string),
		Error: &ErrorInfo{
			Code:    code,
			Message: message,
		},
	}
}

// WithHeader adds a header to the response and returns the response for chaining
func (r *CmdResp) WithHeader(key, value string) *CmdResp {
	if r.Headers == nil {
		r.Headers = make(map[string]string)
	}
	r.Headers[key] = value
	return r
}

// GetHeader returns a header value from the response
func (r *CmdResp) GetHeader(key string) string {
	if r.Headers == nil {
		return ""
	}
	return r.Headers[key]
}

// HasError returns true if the response contains an error
func (r *CmdResp) HasError() bool {
	return r.Error != nil
}

// ContextWithCmdReq embeds CmdReq into a context.Context
func ContextWithCmdReq(ctx context.Context, req *CmdReq) context.Context {
	return context.WithValue(ctx, cmdReqContextKey{}, req)
}

// CmdReqFromContext extracts CmdReq from context.Context
// Returns nil if not present
func CmdReqFromContext(ctx context.Context) *CmdReq {
	if req, ok := ctx.Value(cmdReqContextKey{}).(*CmdReq); ok {
		return req
	}
	return nil
}

// ContextWithCmdResp embeds CmdResp into a context.Context
func ContextWithCmdResp(ctx context.Context, resp *CmdResp) context.Context {
	return context.WithValue(ctx, cmdRespContextKey{}, resp)
}

// CmdRespFromContext extracts CmdResp from context.Context
// Returns nil if not present
func CmdRespFromContext(ctx context.Context) *CmdResp {
	if resp, ok := ctx.Value(cmdRespContextKey{}).(*CmdResp); ok {
		return resp
	}
	return nil
}

// GetRequestHeader is a convenience function to get a request header from context
func GetRequestHeader(ctx context.Context, key string) string {
	if req := CmdReqFromContext(ctx); req != nil {
		return req.Headers[key]
	}
	return ""
}

// SetResponseHeader is a convenience function to set a response header in context
// It creates a CmdResp in context if not present
func SetResponseHeader(ctx context.Context, key, value string) context.Context {
	resp := CmdRespFromContext(ctx)
	if resp == nil {
		resp = &CmdResp{Headers: make(map[string]string)}
		ctx = ContextWithCmdResp(ctx, resp)
	}
	resp.Headers[key] = value
	return ctx
}
